### PR TITLE
[CBRD-21889] Handle duplicate keys in json_keys

### PR DESCRIPTION
**Pull request by cristighr1995**
http://jira.cubrid.org/browse/CBRD-21889

Handle duplicate keys in json_keys

When creating the result json array that will contain all the keys for a specific path, we need to add **only unique values**, so we need to keep track of already inserted keys. For that we can use an unordered_set for O(1) find operation.

Suppose the following example:
```
select json_keys('{"a":{"c":1, "d":2},"a":3, "c":2}');
+------------------------------------------------+
| json_keys('{"a":{"c":1, "d":2},"a":3, "c":2}') |
+------------------------------------------------+
| ["a", "c"]                                     |
+------------------------------------------------+
```
We have 2 keys with name "a" but only one in the result.

### CODE REVIEW

**Comment by eseokoh** [comment]
@cristighr1995 
circleci test is cancelled since there are many crashes of client. 
please investigate it.

File: src/compat/db_json.cpp [hunk 1]
```diff
@@ -708,6 +743,57 @@ db_json_add_element_to_array (JSON_DOC *doc, const JSON_DOC *value)
...
+	    {
+	      const std::string &key = it->name.GetString();
+	      if (inserted_keys.find (key) != inserted_keys.end())
+		{
+		  // we found a duplicate key
+		  duplicate_key = key;
+		  return true;
+		}
+
+	      inserted_keys.insert (key);
```
Comment (paulgheorghecristian) [inline src/compat/db_json.cpp]: I believe insert makes an unnecessary copy of key. how about calling with std::move (key) ?

File: src/compat/db_json.cpp [hunk 2]
```diff
@@ -710,11 +745,63 @@ db_json_add_element_to_array (JSON_DOC *doc, const JSON_DOC *value)
   doc->PushBack (new_doc, doc->GetAllocator ());
 }
 
+static bool
+db_json_contains_duplicate_keys (const JSON_DOC &doc, std::string &duplicate_key)
```
Comment (ribram) [inline src/compat/db_json.cpp]: Use rapidjson Accept () function to map duplicate key check recursively.

File: src/compat/db_json.cpp [hunk 3]
```diff
@@ -1945,7 +2042,7 @@ db_json_get_all_paths_func (const JSON_DOC &doc, JSON_DOC *&result_json)
  * raw_path (in)           : specified path
  */
 int
-db_json_keys_func (const JSON_DOC &doc, JSON_DOC &result_json, const char *raw_path)
+db_json_keys_func (const JSON_DOC &doc, JSON_DOC *&result_json, const char *raw_path)
```
Comment (ribram) [inline src/compat/db_json.cpp]: Please explain the signature change. I see no reason for it.

**Comment by ribram** [comment]
@cristighr1995 

There are 26 regressions. Are they expected or should they be fixed?

**Comment by ribram** [comment]
@cristighr1995 

You were right that Accept/handler is not exactly right for duplicate key checker. I added a different class JSON_WALKER that can call functions before and after walking to children in JSON object/arrays. Please use this for your duplicate key checker.

File: src/compat/db_json.cpp [hunk 4]
```diff
@@ -836,28 +1021,31 @@ db_json_insert_func (const JSON_DOC *doc_to_be_inserted, JSON_DOC &doc_destinati
...
-	  // since PushBack does not guarantee its argument is not modified, we are forced to copy here. Hopefully,
-	  // it doesn't do another copy inside.
-	  JSON_VALUE copy_to_be_ins (db_json_doc_to_value (*doc_to_be_inserted), doc_destination.GetAllocator ());
-	  json_parent_p->PushBack (copy_to_be_ins, doc_destination.GetAllocator ());
-	}
-      else
-	{
-	  db_json_value_wrap_as_array (*json_parent_p, doc_destination.GetAllocator ());
+      // we cannot insert to a non existing path
+      return db_json_er_set_path_does_not_exist (json_pointer_string.substr (0, found), &doc_destination);
```
Comment (eseokoh) [inline src/compat/db_json.cpp]: please refer to my remark on #996

**Pull request merged**